tests/h/services/user_signup_test.py

```diff
@@ -62,6 +71,28 @@ def test_signup_passes_through_privacy_acceptance(self, svc):
 
         assert user.privacy_accepted == now
 
+    def test_signup_sets_provided_user_identities(self, svc):
+        identity_data = [{
+            'provider': 'someprovider',
+            'provider_unique_id': 1
+        }, {
+            'provider': 'someotherprovider',
+            'provider_unique_id': '394ffa3'
+        }]
+
+        user = svc.signup(username='foo',
+                          email='foo@bar.com',
+                          identities=identity_data)
+
+        assert len(user.identities) == 2
+
+    def test_signup_raises_with_invalid_identities(self, svc, db_session):
```

Comment:
This is arguably not a test of the service itself but the underlying models. Let me know if it's inappropriate.

Reply:
I think there's an issue here but I'm not sure what the correct solution is. I need to investigate our other services and views to see how its been solved elsewhere.

The thing is, `signup()` _doesn't_ always raise `IntegrityError`. The `IntegrityError` is raised when the database session gets flushed, if invalid objects have been added to the session. (When the db session is flushed is when sqlalchemy sends the actual SQL commands to postgres for all the objects that've been added to the session, or modified or deleted, so this is when Postgres might raise an error which in turns causes sqlalchemy to raise `IntegrityError`.) `signup()` doesn't always call `self.session.flush()`.

In this test, if `require_activation=True` then `signup()` calls `_require_activation()` and `_require_activation()` does call `flush()` which raises. (`_require_activation()` contains an explicit `flush()` call for some reason.)

But in real usage of the user create API `require_activation=False` is passed to `signup()` and it _doesn't_ call `_require_activation()`, doesn't flush the db session, and doesn't raise. If you change the test to pass `require_activation=False` to `signup()` then the test fails because no exception is raised.

At some point later in the request processing, after `signup()` has finished executing, the db session does get committed and at that point `IntegrityError` is raised and we send back a 500 Server Error:

```python
>>> requests.post("http://localhost:5000/api/users", auth=("...", "..."), data=json.dumps(dict(..., identities=[{<DUPLICATE_IDENTITY>}]))).reason
'Internal Server Error'
```

Reply:
My first thought is that `signup()` have an explicit call to `flush()` after adding the potentially invalid `UserIdentity` objects to the session, and it should have tests like the one you've added here to make sure that it raises `IntegrityError`. Because if `signup()` adds invalid data to the session and doesn't flush then its "poisoned" the session, the session will raise an `IntegrityError` at some later point, either near the end of request processing when Pyramid tries to commit the transaction, or potentially in any other part of the code that calls `flush()` or does a DB query (triggering a sqlalchemy autoflush) and gets an unexpected `IntegrityError` because of `signup()`'s previous bad behaviour.

That's probably the answer, and probably applies to all model and service methods that add potentially invalid data to the session, but I'm not sure how I feel about peppering `flush()` calls throughout our codebase, is that an okay thing to do or do we prefer to keep them to a minimum so that, as much as possible, everything is flushed all at once when the transaction is committed at the end of request processing? That's what I want to look into a little. But I think "flush early and flush often" is probably gonna turn out to be the answer.

Reply:
Another alternative is that we add potentially invalid things to the session as much as we want, never flush, and let the session raise `IntegrityError` when Pyramid tries to commit the transaction (or, potentially, when any part of our code calls `flush()` or tries to do a DB query and triggers and autoflush) but have an exception view or something higher up that catches any `IntegrityError` thrown from anywhere and turns it into a proper error response instead of a 500.